### PR TITLE
Introduced a new arguement --skymap-only-ifos in pycbc_live

### PR DESCRIPTION
**Pull request by SouradeepPal**
A new argument ```--skymap-only-ifos``` for ```pycbc_live``` that accepts a list of IFOs to be used in skymap-only followup. These ifos should now be skipped during the regular matched-filtering, coincidence making, background estimation..

### CODE REVIEW

**Comment by titodalcanton** [comment]
Thanks @SouradeepPal. It looks like this is making the PyCBC Live test fail. Could you have a look?

File: bin/pycbc_live
Comment (SouradeepPal) [reply bin/pycbc_live]: Hi Gareth, all the changes you suggested have been made.

Comment (GarethCabournDavies) [reply bin/pycbc_live]: Thanks! I'll let @titodalcanton hit the approve button, in case there was anything else I missed

File: bin/pycbc_live
Comment (titodalcanton) [reply bin/pycbc_live]: No, that is not a good idea. What @SouradeepPal did initially is the recommended approach.

File: bin/pycbc_live
Comment (titodalcanton) [inline bin/pycbc_live:917]: Do not set defaults here.

File: bin/pycbc_live
Comment (titodalcanton) [inline bin/pycbc_live:1111]: This logic is not really what we want, because it will print "Insufficient data for %s analysis" even if a detector is a skymap-only detector.

File: bin/pycbc_live
Comment (titodalcanton) [inline bin/pycbc_live:1175]: I think you can simplify the code if you set `skymap_only_ifos` to be an attribute of the `LiveEventManager` class. Then you do not need to carry around all these arguments.

File: bin/pycbc_live
Comment (titodalcanton) [reply bin/pycbc_live]: For reference: https://www.quora.com/In-Python-Why-shouldnt-the-default-argument-be-to-make-an-empty-list

File: bin/pycbc_live
Comment (SouradeepPal) [reply bin/pycbc_live]: This is what I had initially thought, but dropped the idea seeeing that ```ifos``` is not an attribute of the ```LiveEventManager```. Is there  a reason. Now I have added ```skymap_only_ifos``` as an attribute which seems to have simplified things. But the [test](https://github.com/gwastro/pycbc/actions/runs/4884286183/jobs/8716818008#step:4:47) failed which I cannot understand atm. Where should I look at to debug.

Comment (GarethCabournDavies) [reply bin/pycbc_live]: The test failure seems to be a condor installation issue - I will click to restart failed tests in case this was a transient thing, but we may need a fix elsewhere if it isnt

**Comment by titodalcanton** [comment]
No, I would like to see at least a minimal proof that this works correctly before merging.

**Comment by SouradeepPal** [comment]
I am retrying but last time I had to stop here: https://git.ligo.org/pycbc/action-items/-/issues/23#note_712046

**Comment by spxiwh** [comment]
@SouradeepPal Checking I've understood: The second row of plots are both from the standard HLV analyses, and we think that one of these is the SNR optimized event?

The third row is from the new HLv analyses, again we think that one of these is SNR optimized, but worry it may not have used V? Can you point me to the trigger directory for both triggers in the bottom row?

**Comment by SouradeepPal** [comment]
Yes, that is what I am guessing. Triggers for HLv are here: ```/home/souradeep.pal/pycbc/examples/live/output_HLv/2020_05_06/candidate_1272790260.113_xagp``` and ```/home/souradeep.pal/pycbc/examples/live/output_HLv/2020_05_06/candidate_1272790260.113_bglu```. I noticed that the later actually contains a single. Please let me know if you can access it.

**Comment by spxiwh** [comment]
Okay, it seems that one of these triggers is a H1 single, which included L1 as a "followup detector" and a H1L1 double, which included V1 as a followup detector. So question is, why did the singles analysis not include V1.

The optimize_snr is done in a subdirectory (of the H1L1 double trigger) called `optimize_snr`. This seems to have failed, but I'm not sure why. We should understand why this failure occurred (unless it's a random weird one-off thing) as well as fixing the single.

File: bin/pycbc_live
Comment (GarethCabournDavies) [inline bin/pycbc_live:462]: I think this also needs to be done in the singles case at around line 547

Comment (spxiwh) [reply bin/pycbc_live:462]: I was just about to write that!

**Comment by SouradeepPal** [comment]
Thanks, HLv skymaps now:
HLv2new

which is now identical to the HLV case.

The ```optimize_snr``` has a log file: 
```Traceback (most recent call last):
  File "/home/souradeep.pal/.conda/envs/pycbc-live-dev/bin/pycbc_optimize_snr", line 351, in <module>
    opt_params = optimize_func(bounds, args, extra_args)
  File "/home/souradeep.pal/.conda/envs/pycbc-live-dev/bin/pycbc_optimize_snr", line 133, in optimize_di
    results = differential_evolution(
  File "/home/souradeep.pal/.conda/envs/pycbc-live-dev/lib/python3.9/site-packages/scipy/optimize/_differentialevolution.py", line 329, in differential_evolution
    ret = solver.solve()
  File "/home/souradeep.pal/.conda/envs/pycbc-live-dev/lib/python3.9/site-packages/scipy/optimize/_differentialevolution.py", line 903, in solve
    result = minimize(self.func,
  File "/home/souradeep.pal/.conda/envs/pycbc-live-dev/lib/python3.9/site-packages/scipy/optimize/_minimize.py", line 692, in minimize
    res = _minimize_lbfgsb(fun, x0, args, jac, bounds,
  File "/home/souradeep.pal/.conda/envs/pycbc-live-dev/lib/python3.9/site-packages/scipy/optimize/_lbfgsb_py.py", line 296, in _minimize_lbfgsb
    raise ValueError("LBFGSB - one of the lower bounds is greater than an upper bound.")
ValueError: LBFGSB - one of the lower bounds is greater than an upper bound.
```

Interestingly, the same error log is also present in one of the HLV triggers.. while another HLV trigger seems to have successfully run the optimized thing.. ```/home/souradeep.pal/pycbc/examples/live/output_HLV/2020_05_06``` The error log also seems to be related to some other thing..

**Comment by spxiwh** [comment]
I would like to be sure the the optimize_snr failure is not caused by this patch. Is it possible to try running the example (HLV) with this patch *not* applied, and see if we can get the same failure?

**Comment by SouradeepPal** [comment]
I meant that HLV is **without** this patch, had the error in one case while not in some other

**Comment by SouradeepPal** [comment]
Sorry, i can check

**Comment by spxiwh** [comment]
The XLAL Errors are just warnings in this case, you should be able to see some form of output from the code in this case??

**Comment by SouradeepPal** [comment]
No, all these got dumped in the log file, no other files were generated

**Comment by spxiwh** [comment]
Well `/home/souradeep.pal/pycbc/examples/live/output/2020_05_06/candidate_1272790260.113_zqdh` looks like it failed, and seems to have evolved quite differently from:

`/home/souradeep.pal/dump/pycbc_new/examples/live/output/2020_05_06/candidate_1272790100.164_aimn`

However, the times are different, so it feels like I'm not comparing like-for-like here(??)

**Comment by SouradeepPal** [comment]
Yes, the two triggers are different, but it is not clear why ```/home/souradeep.pal/dump/pycbc_new/examples/live/output/2020_05_06/candidate_1272790100.164_aimn``` did not generate output files or timeouts do not write logs..

**Comment by spxiwh** [comment]
@SouradeepPal Well there was a log for `candidate_1272790100.164_aimn` but it stops abruptly around 400 seconds after it started. This is consistent with it having been terminated because of the hard timeout that this job is run with. ... We could perhaps try something more fancy to get this event to be logged, but PyCBC Live, in production, must avoid the case that these optimizer jobs don't finish and interfere with operations.

... It does seem like the injection at 1272790100.1 has some "interesting" mass parameters, and the optimizer may well fail in this case without some of the patches being proposed, I just want to check that *this patch* doesn't introduce a failure for this event .... So it would be good to see an snr_optimizer log for the 1272790100.1 trigger before this patch is applied.

**Comment by SouradeepPal** [comment]
But here ```/home/souradeep.pal/dump/pycbc_new/examples/live/output/2020_05_06/candidate_1272790100.164_aimn/optimize_snr``` is from running in a fresh ```e400a51489c3b6b248e7f71393fe7305a9ef0b9c``` from where this branch was created.. note the new path (at ```/home/souradeep.pal/dump/pycbc_new```) where I did a standard run after a fresh install. Sorry, if I am again missing something

**Comment by spxiwh** [comment]
@SouradeepPal Can you be more specific, I could parse "fresh e400a51489c3b6b248e7f71393fe7305a9ef0b9c from where this branch was created" in a few different ways. 

.... But the bottom line is that I want to be convinced that this patch is not introducing a failure to optimize_snr (which could happen if, for e.g., it's input files were not quite being generated correctly). Maybe I confused myself earlier and the case to focus on is 1272790260 where the failure occurs. Does this still occur without this patch applied??

**Comment by SouradeepPal** [comment]
A few more tries **without** this patch applied (or on the commit above) seem to not fail like that with the patch applied (but have other XLAL Error and timeouts)..  So I wanted to understand what might be causing it, here it seems that the process starts, continues for ~ 2min and then fails due to something- so likely not that any missing/misbehaving input file? (I am again pasting the complete log for 1272790260 here at ```/home/souradeep.pal/pycbc/examples/live/output_HLV/2020_05_06/candidate_1272790260.106_pwkj/optimize_snr```)

```
2023-06-25T23:18:28.686-07:00 Running with CPU support: 1 threads
2023-06-25T23:18:30.363-07:00 Starting optimize SNR
2023-06-25T23:18:30.396-07:00 Starting optimization
2023-06-25T23:18:30.590-07:00 Currently at 0 0.0
2023-06-25T23:18:30.741-07:00 Currently at 1 0.0
2023-06-25T23:18:30.900-07:00 Currently at 2 0.0
2023-06-25T23:18:31.027-07:00 Currently at 3 0.0
2023-06-25T23:18:31.210-07:00 Currently at 4 0.0
2023-06-25T23:18:31.446-07:00 Currently at 5 0.0
2023-06-25T23:18:31.664-07:00 Currently at 6 0.0
2023-06-25T23:18:31.938-07:00 Currently at 7 0.0
2023-06-25T23:18:32.250-07:00 Currently at 8 0.0
2023-06-25T23:18:32.634-07:00 Currently at 9 0.0
2023-06-25T23:18:33.169-07:00 Currently at 10 0.0
2023-06-25T23:18:33.779-07:00 Currently at 11 0.0
2023-06-25T23:18:34.593-07:00 Currently at 12 0.0
2023-06-25T23:18:35.707-07:00 Currently at 13 0.0
2023-06-25T23:18:37.226-07:00 Currently at 14 0.0
2023-06-25T23:18:39.160-07:00 Currently at 15 0.0
2023-06-25T23:18:41.566-07:00 Currently at 16 0.0
2023-06-25T23:18:43.681-07:00 Currently at 17 0.0
2023-06-25T23:18:45.647-07:00 Currently at 18 0.0
2023-06-25T23:18:48.511-07:00 Currently at 19 0.0
2023-06-25T23:18:51.607-07:00 Currently at 20 0.0
2023-06-25T23:18:54.406-07:00 Currently at 21 0.0
2023-06-25T23:18:57.274-07:00 Currently at 22 0.12369231656981065
2023-06-25T23:19:00.154-07:00 Currently at 23 0.12158622441614034
2023-06-25T23:19:03.603-07:00 Currently at 24 0.12579705866462773
2023-06-25T23:19:06.906-07:00 Currently at 25 0.1276872044443044
2023-06-25T23:19:10.011-07:00 Currently at 26 0.12161535414374221
2023-06-25T23:19:12.753-07:00 Currently at 27 0.12120462574336378
2023-06-25T23:19:15.714-07:00 Currently at 28 0.11954771776250084
2023-06-25T23:19:19.262-07:00 Currently at 29 0.11552664213923491
2023-06-25T23:19:22.230-07:00 Currently at 30 0.11528027571920592
2023-06-25T23:19:25.223-07:00 Currently at 31 0.11576148588313943
2023-06-25T23:19:28.178-07:00 Currently at 32 0.11615297295499952
2023-06-25T23:19:31.552-07:00 Currently at 33 0.10980472141997731
2023-06-25T23:19:35.004-07:00 Currently at 34 0.10918058644014089
2023-06-25T23:19:38.167-07:00 Currently at 35 0.10951318911761426
2023-06-25T23:19:41.177-07:00 Currently at 36 0.10991558238299652
2023-06-25T23:19:44.521-07:00 Currently at 37 0.10935057623251014
2023-06-25T23:19:48.182-07:00 Currently at 38 0.11175375326354606
2023-06-25T23:19:51.273-07:00 Currently at 39 0.10937568390898102
2023-06-25T23:19:54.937-07:00 Currently at 40 0.1046167384450629
2023-06-25T23:19:58.868-07:00 Currently at 41 0.10293193408911076
2023-06-25T23:20:02.090-07:00 Currently at 42 0.10397934569719869
2023-06-25T23:20:05.029-07:00 Currently at 43 0.10542646896213599
2023-06-25T23:20:08.264-07:00 Currently at 44 0.10676131040337074
2023-06-25T23:20:11.990-07:00 Currently at 45 0.10741358539407901
2023-06-25T23:20:15.553-07:00 Currently at 46 0.10714211281577206
2023-06-25T23:20:18.890-07:00 Currently at 47 0.10949187235934509
2023-06-25T23:20:22.356-07:00 Currently at 48 0.11203763171188066
2023-06-25T23:20:26.088-07:00 Currently at 49 0.11555929608029122
Traceback (most recent call last):
  File "/home/souradeep.pal/.conda/envs/pycbc-live-dev/bin/pycbc_optimize_snr", line 351, in <module>
    opt_params = optimize_func(bounds, args, extra_args)
  File "/home/souradeep.pal/.conda/envs/pycbc-live-dev/bin/pycbc_optimize_snr", line 133, in optimize_di
    results = differential_evolution(
  File "/home/souradeep.pal/.conda/envs/pycbc-live-dev/lib/python3.9/site-packages/scipy/optimize/_differentialevolution.py", line 329, in differential_evolution
    ret = solver.solve()
  File "/home/souradeep.pal/.conda/envs/pycbc-live-dev/lib/python3.9/site-packages/scipy/optimize/_differentialevolution.py", line 903, in solve
    result = minimize(self.func,
  File "/home/souradeep.pal/.conda/envs/pycbc-live-dev/lib/python3.9/site-packages/scipy/optimize/_minimize.py", line 692, in minimize
    res = _minimize_lbfgsb(fun, x0, args, jac, bounds,
  File "/home/souradeep.pal/.conda/envs/pycbc-live-dev/lib/python3.9/site-packages/scipy/optimize/_lbfgsb_py.py", line 296, in _minimize_lbfgsb
    raise ValueError("LBFGSB - one of the lower bounds is greater than an upper bound.")
ValueError: LBFGSB - one of the lower bounds is greater than an upper bound.
```

Also wanted to flag slightly different error also occured in one of the tests, at ```/home/souradeep.pal/dump/pycbc_new/examples/live/output_01/2020_05_06/candidate_1272790260.106_byqa/optimize_snr```
```
2023-06-27T10:54:43.351-07:00 Running with CPU support: 1 threads
2023-06-27T10:54:44.878-07:00 Starting optimize SNR
2023-06-27T10:54:44.913-07:00 Starting optimization
2023-06-27T10:54:45.151-07:00 Currently at 0 0.0
2023-06-27T10:54:45.323-07:00 Currently at 1 0.0
2023-06-27T10:54:45.489-07:00 Currently at 2 0.0
2023-06-27T10:54:45.697-07:00 Currently at 3 0.0
2023-06-27T10:54:45.959-07:00 Currently at 4 0.0
2023-06-27T10:54:46.151-07:00 Currently at 5 0.0
2023-06-27T10:54:46.436-07:00 Currently at 6 0.0
2023-06-27T10:54:46.767-07:00 Currently at 7 0.0
2023-06-27T10:54:47.180-07:00 Currently at 8 0.0
2023-06-27T10:54:47.619-07:00 Currently at 9 0.0
2023-06-27T10:54:48.089-07:00 Currently at 10 0.0
2023-06-27T10:54:48.637-07:00 Currently at 11 0.0
2023-06-27T10:54:49.459-07:00 Currently at 12 0.0
2023-06-27T10:54:50.696-07:00 Currently at 13 0.0
2023-06-27T10:54:52.565-07:00 Currently at 14 0.0
2023-06-27T10:54:54.895-07:00 Currently at 15 0.0
2023-06-27T10:54:57.668-07:00 Currently at 16 0.0
2023-06-27T10:55:00.339-07:00 Currently at 17 0.0
2023-06-27T10:55:03.424-07:00 Currently at 18 0.15462157602934376
2023-06-27T10:55:06.209-07:00 Currently at 19 0.16560150890361205
2023-06-27T10:55:09.992-07:00 Currently at 20 0.17236986544963082
2023-06-27T10:55:13.146-07:00 Currently at 21 0.17882036802081058
2023-06-27T10:55:15.927-07:00 Currently at 22 0.1865365689618426
2023-06-27T10:55:19.042-07:00 Currently at 23 0.19182247229642543
2023-06-27T10:55:22.343-07:00 Currently at 24 0.19405987525030838
2023-06-27T10:55:25.522-07:00 Currently at 25 0.20203903396052553
2023-06-27T10:55:28.662-07:00 Currently at 26 0.20411059874279117
2023-06-27T10:55:31.873-07:00 Currently at 27 0.20520980126113925
2023-06-27T10:55:34.528-07:00 Currently at 28 0.20420279829960586
2023-06-27T10:55:37.400-07:00 Currently at 29 0.20281260104309384
2023-06-27T10:55:40.478-07:00 Currently at 30 0.20073743632326843
2023-06-27T10:55:43.343-07:00 Currently at 31 0.20419028268475006
2023-06-27T10:55:46.111-07:00 Currently at 32 0.19804654985944953
2023-06-27T10:55:48.537-07:00 Currently at 33 0.20253832569744024
2023-06-27T10:55:50.922-07:00 Currently at 34 0.19518764900428548
2023-06-27T10:55:53.440-07:00 Currently at 35 0.18661245489467423
2023-06-27T10:55:55.938-07:00 Currently at 36 0.18725025554715025
2023-06-27T10:55:58.781-07:00 Currently at 37 0.19005949592404217
2023-06-27T10:56:02.166-07:00 Currently at 38 0.19002199548402401
2023-06-27T10:56:04.677-07:00 Currently at 39 0.18468736075098843
2023-06-27T10:56:07.380-07:00 Currently at 40 0.17153755680413577
2023-06-27T10:56:10.209-07:00 Currently at 41 0.1700399314608005
2023-06-27T10:56:12.938-07:00 Currently at 42 0.16439953546043698
2023-06-27T10:56:15.565-07:00 Currently at 43 0.15151417248297913
2023-06-27T10:56:18.492-07:00 Currently at 44 0.15189149977458652
2023-06-27T10:56:21.146-07:00 Currently at 45 0.1452060009947811
2023-06-27T10:56:23.687-07:00 Currently at 46 0.14563697856653474
2023-06-27T10:56:26.522-07:00 Currently at 47 0.13801852087973493
2023-06-27T10:56:29.530-07:00 Currently at 48 0.13521316039258574
2023-06-27T10:56:32.129-07:00 Currently at 49 0.13278678595439428
Traceback (most recent call last):
  File "/home/souradeep.pal/.conda/envs/pycbc-live-ori/bin/pycbc_optimize_snr", line 351, in <module>
    opt_params = optimize_func(bounds, args, extra_args)
  File "/home/souradeep.pal/.conda/envs/pycbc-live-ori/bin/pycbc_optimize_snr", line 133, in optimize_di
    results = differential_evolution(
  File "/home/souradeep.pal/.conda/envs/pycbc-live-ori/lib/python3.9/site-packages/scipy/optimize/_differentialevolution.py", line 402, in differential_evolution
    ret = solver.solve()
  File "/home/souradeep.pal/.conda/envs/pycbc-live-ori/lib/python3.9/site-packages/scipy/optimize/_differentialevolution.py", line 1082, in solve
    result = minimize(self.func,
  File "/home/souradeep.pal/.conda/envs/pycbc-live-ori/lib/python3.9/site-packages/scipy/optimize/_minimize.py", line 643, in minimize
    bounds = _validate_bounds(bounds, x0, meth)
  File "/home/souradeep.pal/.conda/envs/pycbc-live-ori/lib/python3.9/site-packages/scipy/optimize/_minimize.py", line 996, in _validate_bounds
    raise ValueError(msg)
ValueError: An upper bound is less than the corresponding lower bound.
```

**Comment by GarethCabournDavies** [comment]
Could you rebase to master? (There should be a 'sync fork' option when you go to https://github.com/SouradeepPal/pycbc/tree/skymap-only)

I think this could be solved by #4338, but I'm not 100% (and not sure that is missing from this PR's branch)

**Comment by SouradeepPal** [comment]
> Could you rebase to master? ~(There should be a 'sync fork' option when you go to https://github.com/SouradeepPal/pycbc/tree/skymap-only)~ Actually it would be easier for you to do this locally
> 
> I think this could be solved by #4338, but I'm not 100%

**Comment by SouradeepPal** [comment]
I had synced fork, now I accidentally closed the pr!

**Comment by spxiwh** [comment]
I guess my only concern with this is ensuring that the optimizer gets *the right* data for V1 in the case of V1 being used as a skymap-only ifo. Maybe this could be checked in different ways (e.g. comparing data with/without the skymap-only-ifo flag that the optimizer gets sent).I do agree that the optimizer before the patch Gareth mentions may never have worked well in the cases here.

File: pycbc/io/live.py
Comment (spxiwh) [inline pycbc/io/live.py:95]: Please delete this line (and the comment line above)

File: pycbc/io/live.py
Comment (spxiwh) [inline pycbc/io/live.py:104]: And this line

File: pycbc/io/live.py
Comment (spxiwh) [inline pycbc/io/live.py:198]: And all of these lines now commented out.

**Review by spxiwh — APPROVED** [review]
@SouradeepPal Okay thanks. I think this is now ready to move over to testing in the MDC (@titodalcanton). I'm giving approval to merge, but please remove the commented lines first as requested (this should also restart the CI tests), then we can merge.

File: pycbc/io/live.py
Comment (spxiwh) [inline pycbc/io/live.py:196]: `bayestar_check_fields` and the comment above should also now be removed, as it is now unused.

**Comment by tdent** [comment]
@SouradeepPal to deal with the question of why things are reproducible or identical, we suggested on the call to give the random seed option in optimize_snr a default value (eg 42 ..) so that it's always used unless someone overrides it.

**Comment by spxiwh** [comment]
@tdent I was basically ready to press the merge button on this. Do you want to hold up to make that change ... or not?

**Comment by tdent** [comment]
We had discussion on the pycbc call and from what I remember I don't know if Tito was convinced by the testing and it didn't seem like there was a consensus that this should be merged immediately. 
Including a fixed random seed for reproducibility may help to address that.

**Comment by tdent** [comment]
FWIW, as far as I can see the previous optimizer errors should be addressed by the recent commits that fix the mass ranges to be 'saner'.

**Comment by SouradeepPal** [comment]
It seems fixing the ```seed``` in ```pycbc_optimize_snr``` produces the same skymap in the optimized HLV and HLv cases:
seeded_HLVLe_HLvRi

The contents of the ```coinc*.xml``` files now look identical, the snrseries, psds.. have run >3 times and each produced the same results. Also I did not face any error with optimization in these tests.

So I have pushed a default of 42 for the seed arg in the optimize_snr code. Is that okay.

**Comment by spxiwh** [comment]
Okay, I'll merge this now as soon as I can persuade the CI tests to pass.

**Comment by spxiwh** [comment]
@SouradeepPal Unfortunately this needs to be rebased again. :-(

File: bin/pycbc_live
Comment (SouradeepPal) [reply bin/pycbc_live:469]: Yes, it was done probably [here](https://github.com/gwastro/pycbc/pull/4346/commits/91ee29ff2c20fa18ee88f01fd01ad5b4fca883e2), but seemed to have failed tests

**Pull request merged**
